x/p2p/client.go

```diff
@@ -0,0 +1,183 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"golang.org/x/exp/maps"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/utils/sampler"
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+var (
+	ErrAppRequestFailed = errors.New("app request failed")
+	ErrRequestPending   = errors.New("request pending")
+	ErrNoPeers          = errors.New("no peers")
+)
+
+// AppResponseCallback is called upon receiving an AppResponse for an AppRequest
+// issued by Client.
+// Callers should check [err] to see whether the AppRequest failed or not.
+type AppResponseCallback func(
+	nodeID ids.NodeID,
+	responseBytes []byte,
+	err error,
+)
+
+// CrossChainAppResponseCallback is called upon receiving an
+// CrossChainAppResponse for a CrossChainAppRequest issued by Client.
+// Callers should check [err] to see whether the AppRequest failed or not.
+type CrossChainAppResponseCallback func(
+	chainID ids.ID,
+	responseBytes []byte,
+	err error,
+)
```

Comment:
Would we ever want to allow the VM to return a fatal error to the consensus engine from here?

Reply:
Chatted about this offline but we don't think this will be a common use-case. If we really need a vm to fatal at any point we can add a hook into snow context.